### ISSUE TITLE
Import Module from Template

### ISSUE THREAD

**Issue by R-Bower** [angular/vscode-ng-language-service#1917]
This ticket was closed despite not having been resolved.

              I'm intending to enable this for external libraries in a future release. I was worried about performance issues, and so I put it on the back burner for a while. (Although the primary motivation is standalone, there's no reason we can't also do it for NgModules)

_Originally posted by @dylhunn in https://github.com/angular/vscode-ng-language-service/issues/1501#issuecomment-1520730889_

I'm posting it again here because this is a major blocker for my transition from WebStorm to vscode.  It may seem minor, but my productivity is impacted by this issue.

# 🚀 feature request

### Description

One of the killer features of Webstorm when working with angular is the ability to automatically import NgModules from templates as an auto completion for components, directives and pipes.

It would be fantastic if such a feature could be brought into vsc
### Feature Type
What does this bug affect

- [x] Angular Language Service VSCode extension
- [x] Angular Language Service server

### Describe the solution you'd like

This is probably a tricky problem to solve as it relies on knowing all the possible NgModules (something that even TypeScript struggles with for things like submodules imports of @angular/material) and then adding the import via a code completion to a file that is potentially not open.

The closest thing I can think of that currently exist in vsc is renaming files automatically updating their imports which I believe the TypeScript language service manages.

### Describe alternatives you've considered

Use WebStorm instead of vscode.

**Comment by dylhunn** [comment]
This would be a great feature. It is a performance challenge though, because the `DtsMetadataReader` needs to be able to enumerate all the selectors found in all directive/component metas in all Dts files.

**Comment by m4riok** [comment]
Automatic Imports from your own code base make way less sense than imports from an external library. This feature will seriously boost productivity. I too would love to see this implemented in vsc. If performance is an issue perhaps provide an option for loading all / some selectors so the user can choose whether he is willing to suffer the performance penalty in order to have this feature ?

**Comment by Martinspire** [comment]
I'm not entirely famililar with how webstorm does it, but do you propose something like this:
When you use standalone modules, it doesn't recognize various components when you add them in your template. It would be nice if it automatically updated the imports when you use something in the template that isn't defined yet.

So when I would type `<second-component />` then it would automatically update the imports of my component. Same for when I'm using forms that it would automatically update to add FormsModule and stuff like that. And when you removed one, it should also clean it up. Preferrably sort them alphabetically or perhaps modules first and then components and so on.

And for when you aren't using standalone modules, it would be nice if things get imported into the modules file so that you can use it immediately without having to worry about adding it there.

**Comment by m4riok** [comment]
@Martinspire If I am not mistaken Webstorm does it by maintaining it's own indexes which it builds on startup and updates when changes are made to package.json.

Functionality is exactly as you describe it except it does not automatically remove the import when you remove the selector from the template, only adds it automatically when you add it.

**Comment by asphub** [comment]
Need an option to import necessary modules or components directly from the HTML, at least for the _**standalone**_ components. 
Need to show a **Quick Fix** bulb can be used for such imports. 

![322353785-3c6b26b9-12fa-4c28-8a07-b8a2724af606](https://github.com/user-attachments/assets/694277ac-96de-4b48-81d1-59f5ee797b6c)

**Comment by nuriddin2468** [comment]
is there any news on this ?

**Comment by codetroll** [comment]
Yeah - this is sorely lacking in VS Code. As it is now, you don't even see that there is something missing.

**Comment by gnibeda** [comment]
Any news or workaround? 
How you guys working now with Angular 19 where all components standalone by default? 
It is nightmare manually import each directive and component when writing templates.

**Comment by JWess** [comment]
This works great for me if I'm importing my own standalone components from my project. But it doesn't work at all when I'm trying to import a component from `node_modules`. Agreed with those above...it's so painful to have to manually import tens of components/pipes/directives every time I create a new component. When working with Angular Material, it's not uncommon to import 10 different Material components in one template file. Right now, that means manually adding the following to my component.ts file, for example:

```
MatAutocomplete,
MatAutocompleteTrigger,
MatButton,
MatDatepicker,
MatDatepickerInput,
MatDatepickerToggle,
MatFormField,
MatIcon,
MatInput,
MatLabel,
MatOption,
MatProgressBar,
MatSelect,
MatSuffix,
MatTooltip,
```

**Comment by R-Bower** [comment]
> How are you guys working now with Angular 19, where all components are standalone by default?

WebStorm

**Comment by sonderbazz** [comment]
[VScode is the most popular IDE for developers](https://survey.stackoverflow.co/2024/technology?ref=runtime.news#1-integrated-development-environment) - so why shouldn't it support auto imports within the template file? 
It seams like the Angular team is pushing the use of standalone components - so why not help us out doing so.

At very least it should be able to auto import Angular own components and Angular Material components. I perfectly understand that having the Language service index all the packages within node_modules would kill all performance. But surely having the angular package indexed shouldn't be that heavy on performance.

**Comment by Harpush** [comment]
Any news concerning this? Really needed

**Comment by karthickvasudev** [comment]
yes really needed

**Comment by karthickvasudev** [comment]
> > yes really needed
> 
> https://marketplace.visualstudio.com/items?itemName=baryshevrs.angular-auto-import
> 
> This plugin helps import modules.

i have tried this one but not working as expected

<img width="890" height="312" alt="Image" src="https://github.com/user-attachments/assets/2a98f90d-83b0-4b07-9868-73449a0b326f" />

tried reindexing, cache clearing tough luck..

**Comment by D-Biela** [comment]
> The latest Angular extension now supports importing components from `node_modules`. You should give it a try.

Does not work with directives. Everything else except directives seems to work fine though.
